README.markdown
SMSPie
=============================================================================

About
------------
Send SMS messages using the providers website. As of writting it supports 
meteor.ie and will support vodafone and o2 once there are accounts to test with.

Usage
------------

	Usage: smspie 0851234567

	Send SMS via command line

	Options:
	  --version             show program's version number and exit
	  -h, --help            show this help message and exit
	  -C PROVIDER, --provider=PROVIDER
		                sets the SMS provider
	  -c CONFIG, --config=CONFIG
		                sets the path to the config file
	  -i, --import-contacts
		                import Meteor contacts, copy and paste into the config
		                file
	  -l, --list            list the people in your phonebook
	  -m MESSAGE, --message=MESSAGE
		                the message you want to send
	  -u USERNAME, --username=USERNAME
		                set your username
	  -p PASSWORD, --password=PASSWORD
		                set your password
	  -v, --verbose         set the output as verbose (debug)


Installation
------------
To install run this as root: `python setup.py install`

To do a local install do: `python setup.py install --home="$HOME"`

Examples
------------
Importing contacts for your phonebook

	$ smspie -C meteor --import-contacts
	Add the following to the ~/.smspie/config.yaml under the section phonebook

	phonebook:
	- {phonenumber: 0851234567, username: John}
	- {phonenumber: 0851234567, username: Smith}

Sending a message to a person in your phonebook

	$ smspie -C meteor John
	[ Using a session for 0851234567@meteor ... ]
	[ Recipient: John (0851234567) ]
	> oh hi, lets met up later.
	> 
	[ Message sent to John (0851234567) ]

Sending a message to a number

	$ smspie -C meteor 0851234567
	[ Using a session for 0851234567@meteor ... ]
	[ Recipient:  (0851234567) ]
	> oh hi, lets met up later.
	> 
	[ Message sent to  (0851234567) ]

Listing the contacts you have

	$ smspie -l
	John          0851234567
	Smith         0851234567

Extra
------------
Something extra you might like to do is alias the commands, as mentioning what
section your provider is from is tedious. So `alias meteorsms="smspie -C meteor"`
can save you time, and can also be handy for multiple-user systems.


ChangeLog
2009-10-08  d_fens <d_fens@redbrick.dcu.ie>

  * README.markdown, scripts/smspie: Updated README and smspie TODO for names
  and also the comparison method was altered

  * smspie/providers/__init__.py: __init__ files missing

2009-10-07  d_fens <d_fens@redbrick.dcu.ie>

  * scripts/smspie, smspie/providers/http.py, smspie/providers/meteor.py:
  Forgot http class(important), change some py3k bits (probably still not
  compatible), added logging options with -vvv, changed the script to support
  localisation but nothing concrete

2009-10-04  d_fens <d_fens@redbrick.dcu.ie>

  * LICENSE, README.markdown, scripts/smspie, setup.py,
  smspie/providers/meteor.py: Added setup.py, README.markdown, coded the smspie
  script, coded the meteor functionality

2009-08-27  d_fens <d_fens@redbrick.dcu.ie>

  * : Setup structure


smspie/providers/o2.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import os
import time
import re
import cookielib
import pycurl
from BeautifulSoup import BeautifulSoup
import logging

from http import http

class o2(http):
	"""
	"""
	def __init__(self, config={}):
		"""
		Setup class, takes in a config for our settings
		"""
		http.__init__(self)
		self.config = config # TODO: throw warning on required settings 

		self.logger = logging.getLogger(self.__module__)

		if 'proxy' in self.config and 'use_proxy' in self.config and self.config['use_proxy'] and self.config['proxy'] is not None:
			self.curl.setopt(pycurl.PROXY, self.config['proxy'])
		if 'user-agent' in self.config and self.config['user-agent'] is not None:
			self.curl.setopt(pycurl.USERAGENT, self.config['user-agent'])
		if 'cookiefile' in self.config and self.config['cookiefile'] is not None:
			self.curl.setopt(pycurl.COOKIEFILE, os.path.expanduser(self.config['cookiefile']))
			self.curl.setopt(pycurl.COOKIEJAR, os.path.expanduser(self.config['cookiefile']))

	def resumable(self, expiration=3600):
		"""
		"""
		resume = False
		if 'cookiefile' in self.config and self.config['cookiefile'] is not None:
			if os.path.isfile(os.path.expanduser(self.config['cookiefile'])):
				self.logger.debug("Attempting to use cookiefile %s", self.config['cookiefile'])
				try:
					cookiejar = cookielib.MozillaCookieJar(os.path.expanduser(self.config['cookiefile']))
					cookiejar.load()

					cookies = 0
					resume = True # everything loads, but lets check more!
					for index, cookie in enumerate(cookiejar):
						cookies = cookies + 1
						if int(cookie.expires) <= time.time():
							self.logger.info("%s expired %ss ago", cookie, int(cookie.expires - time.time()))
							resume = False
						else:
							self.logger.debug("%s will expire in %ss", cookie, int(cookie.expires - time.time()))
							cookie.expires = int(time.time() + expiration)
					if cookies == 0:
						resume = False

					if not resume:
						cookiejar.clear()

					cookiejar.save()
				except IOError:
					return False
		return resume

	def connect(self):
		"""
		Connects to o2online.ie, does a check to see if o2s up and working.
		"""
		if self.resumable():
			self.logger.info("Resuming from connect.")
			return True

		handle = self.get('http://www.o2online.ie/wps/wcm/connect/O2/Home/')
		soup = BeautifulSoup(handle)

		if unicode(soup.html.head.title.string).strip() == u"Mobile Phones – Prepaid Phones – Bill Pay Phones – Mobile Broadband - Broadband Deals - with O2 Ireland":
			self.logger.info("connect has correct HTML title.")
			return True
		return False

	def login(self, username, password):
		"""
		Login to o2online.ie

		Returns true if successful or false if fails.
		"""
		if self.resumable():
			self.logger.info("Resuming from login.")
			return True
		else:
			self.logger.info("Unable to resume, running connect from login.")
			self.connect()

		post = [
			('IDButton', 'Go'),
			('org', 'o2ext'),
			('CONNECTFORMGET', 'TRUE'),
			('IDToken1', username),
			('IDToken2', password)
		]

		handle = self.post('https://www.o2online.ie/amserver/UI/Login', post)
		from html5lib import HTMLParser, treebuilders
		parser = HTMLParser(tree=treebuilders.getTreeBuilder("beautifulsoup"))
		soup = parser.parse(handle)

		if unicode(soup.html.head.title.string).strip() == u"LoginCheck":
			self.logger.info("login has correct HTML title.")
			return True
		return False

	def message(self, phonenumber, message):
		"""
		Sends a message to the recipient.
		"""
		params = [
			('APIID', 'AUTH-WEBSSO'),
			('TargetApp', 'o2om_smscenter_new.osp?MsgContentID=-1&SID=_'),
			('utm_source', 'dashboard_webtext_link'),
			('utm_medium', 'link'),
			('utm_campaign', 'o2_dashboard'),
		]
		self.get('http://messaging.o2online.ie/ssomanager.osp', params)

		params = [
			('MsgContentID', '-1'),
			('SID', '_'),
			('SID', '8240509_utusnutl')
		]
		handle = self.get('http://messaging.o2online.ie/o2om_smscenter_new.osp', params)

		result = re.search('Number of free text messages remaining this month: <strong>(?P<remaining>\d+)</strong>', handle, re.IGNORECASE)
		if result:
			self.remaning = result.group('remaining')

		from html5lib import HTMLParser, treebuilders
		parser = HTMLParser(tree=treebuilders.getTreeBuilder("beautifulsoup"))
		soup = parser.parse(handle)

		form = soup.find("form", {"name": "frmSMS"})
		if not form:
			self.logger.info("no form")
			return False
		inputs = form.findAll("input", {"type": "Hidden"})
		if not inputs:
			self.logger.info("no inputs")
			return False

		post = []
		for i in inputs:
			post.append((i.get('name'), i.get('value')))
		post.append(('SMSTo', phonenumber))
		post.append(('selcountry', '00355'))
		post.append(('SMSText', message))

		params = [
			('MsgContentID', '-1'),
			('SID', '_'),
			('SID', '8240509_utusnutl')
		]
		handle = self.post('http://messaging.o2online.ie/smscenter_send.osp', post, params)

		return False

	def import_contacts(self):
		"""
		Returns the list of contacts within the phonebook.

		If it fails, there is no checks so it returns an empty list.
		"""
		return []



smspie/providers/http.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import pycurl
import urllib

try:
	from cStringIO import StringIO
except ImportError:
	from cStringIO import StringIO

class http():
	def __init__(self):
		"""
		Setup class, takes in a config for our settings
		"""
		self.buffer = StringIO()
		self.curl = pycurl.Curl()

		self.curl.setopt(pycurl.SSL_VERIFYHOST, 1)
		self.curl.setopt(pycurl.SSL_VERIFYPEER, 0) # TODO: fix

		self.curl.setopt(pycurl.FOLLOWLOCATION, 1)
		self.curl.setopt(pycurl.MAXREDIRS, 9)
		self.curl.setopt(pycurl.CONNECTTIMEOUT, 30)
		self.curl.setopt(pycurl.TIMEOUT, 30)

		# FIXME: broken in certain pycurl versions, fixes vodafone referers
		#self.curl.setopt(pycurl.AUTOREFERER, 1)

		self.curl.setopt(pycurl.VERBOSE, 0)

		self.curl.setopt(pycurl.SSLVERSION, 3)

	def get(self, url, params=None):
		if params:
			url += "?" + urllib.urlencode(params)
		self.curl.setopt(pycurl.URL, url)
		self.curl.setopt(pycurl.HTTPGET, 1)

		self.buffer = StringIO()
		self.curl.setopt(pycurl.WRITEFUNCTION, self.buffer.write)
		self.curl.perform()
		s = self.buffer.getvalue()
		self.buffer.close()
		return s

	def post(self, url, params=None, args=None):
		if params:
			self.curl.setopt(pycurl.POSTFIELDS, urllib.urlencode(params))
		if args:
			url += "?" + urllib.urlencode(args)
		self.curl.setopt(pycurl.URL, url)
		self.curl.setopt(pycurl.POST, 1)

		self.buffer = StringIO()
		self.curl.setopt(pycurl.WRITEFUNCTION, self.buffer.write)
		self.curl.perform()
		s = self.buffer.getvalue()
		self.buffer.close()
		return s

	def close(self):
		self.curl.close()


smspie/providers/vodafone.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import os
import time
import re
import urllib
import cookielib
import pycurl
from BeautifulSoup import BeautifulSoup
import logging

from http import http

class vodafone(http):
	"""
	"""
	def __init__(self, config={}):
		"""
		Setup class, takes in a config for our settings
		"""
		http.__init__(self)
		self.config = config # TODO: throw warning on required settings 

		self.logger = logging.getLogger(self.__module__)

		if 'proxy' in self.config and 'use_proxy' in self.config and self.config['use_proxy'] and self.config['proxy'] is not None:
			self.curl.setopt(pycurl.PROXY, self.config['proxy'])
		if 'user-agent' in self.config and self.config['user-agent'] is not None:
			self.curl.setopt(pycurl.USERAGENT, self.config['user-agent'])
		if 'cookiefile' in self.config and self.config['cookiefile'] is not None:
			self.curl.setopt(pycurl.COOKIEFILE, os.path.expanduser(self.config['cookiefile']))
			self.curl.setopt(pycurl.COOKIEJAR, os.path.expanduser(self.config['cookiefile']))

	def resumable(self, expiration=3600):
		"""
		"""
		resume = False
		if 'cookiefile' in self.config and self.config['cookiefile'] is not None:
			if os.path.isfile(os.path.expanduser(self.config['cookiefile'])):
				self.logger.debug("Attempting to use cookiefile %s", self.config['cookiefile'])
				try:
					cookiejar = cookielib.MozillaCookieJar(os.path.expanduser(self.config['cookiefile']))
					cookiejar.load()

					return True
					cookies = 0
					resume = True # everything loads, but lets check more!
					for index, cookie in enumerate(cookiejar):
						cookies = cookies + 1
						if int(cookie.expires) <= time.time():
							self.logger.info("%s expired %ss ago", cookie, int(cookie.expires - time.time()))
							resume = False
						else:
							self.logger.debug("%s will expire in %ss", cookie, int(cookie.expires - time.time()))
							cookie.expires = int(time.time() + expiration)
					if cookies == 0:
						resume = False

					if not resume:
						cookiejar.clear()

					cookiejar.save()
				except IOError:
					return False
		return resume

	def connect(self):
		"""
		Connects to vodafone.ie, does a check to see if meteors up and working.
		"""
		if self.resumable():
			self.logger.info("Resuming from connect.")
			return True

		handle = self.get('https://www.vodafone.ie/')
		soup = BeautifulSoup(handle)

		if soup.html.head.title.string == "Vodafone Mobile Phone Ireland - Best Mobile Phone &amp; Irish Broadband Deals From Vodafone Ireland":
			self.logger.info("connect has correct HTML title.")
			return True
		return False

	def login(self, username, password):
		"""
		Login to vodafone.ie

		Returns true if successful or false if fails.
		"""
		if self.resumable():
			self.logger.info("Resuming from login.")
			return True
		else:
			self.logger.info("Unable to resume, running connect from login.")
			self.connect()

		post = [
			('username', username),
			('password', password)
		]
		handle = self.post('https://www.vodafone.ie/myv/services/login/Login.shtml', post)
		soup = BeautifulSoup(handle)

		if soup.html.head.title.string == "My Vodafone, Webtext, Email, Upgrades, Vodafone Email Plus, Mobile Phone Offers at Vodafone.ie":
			self.logger.info("login has correct HTML title.")
			return True
		return False

	def message(self, phonenumber, message):
		"""
		Sends a message to the recipient.
		"""
		handle = self.get('https://www.vodafone.ie/myv/messaging/webtext/Setup.shtml')
		soup = BeautifulSoup(handle)
		if soup.html.head.title.string == "Web text - Vodafone Ireland":
			self.logger.info("message has correct HTML title.")

		secret = soup.find("input", {"type": "hidden", "name": "org.apache.struts.taglib.html.TOKEN"})
		if secret and secret.has_key('value'):
			secret = secret.get('value')
			self.logger.info("secret value [%s]", secret)
		else:
			self.logger.warning("no secret value, panic.")
			return False

		post = [
			('org.apache.struts.taglib.html.TOKEN', secret),
			('message', message),
			('recipients[0]', phonenumber),
			('recipients[1]', ''),
			('recipients[2]', ''),
			('recipients[3]', ''),
			('recipients[4]', ''),
			('contact_filter', 'Search contacts'),
			('futuredate', 'false'),
			('futuretime', 'false')
		]
		ref_params = [
			('ts', int(time.time())),
			('keeprecipients', 'true')
		]
		# FIXME: remove if > pycurl becomes more available
		self.curl.setopt(pycurl.REFERER, 'https://www.vodafone.ie/myv/messaging/webtext/Setup.shtml' + urllib.urlencode(ref_params))
		handle = self.post('https://www.vodafone.ie/myv/messaging/webtext/Process.shtml', post)
		soup = BeautifulSoup(handle)

		# TODO: add checking
		return True

	def import_contacts(self):
		"""
		Returns the list of contacts within the phonebook.

		If it fails, there is no checks so it returns an empty list.
		"""
		# TODO: vodafone uses webmail (at least for a nice list of all users)
		return []



smspie/providers/meteor.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import os
import time
import re
import cookielib
import pycurl
from BeautifulSoup import BeautifulSoup
import logging

from http import http

class meteor(http):
	"""
	"""
	def __init__(self, config={}):
		"""
		Setup class, takes in a config for our settings
		"""
		http.__init__(self)
		self.config = config # TODO: throw warning on required settings 

		self.logger = logging.getLogger(self.__module__)

		if 'proxy' in self.config and 'use_proxy' in self.config and self.config['use_proxy'] and self.config['proxy'] is not None:
			self.curl.setopt(pycurl.PROXY, self.config['proxy'])
		if 'user-agent' in self.config and self.config['user-agent'] is not None:
			self.curl.setopt(pycurl.USERAGENT, self.config['user-agent'])
		if 'cookiefile' in self.config and self.config['cookiefile'] is not None:
			self.curl.setopt(pycurl.COOKIEFILE, os.path.expanduser(self.config['cookiefile']))
			self.curl.setopt(pycurl.COOKIEJAR, os.path.expanduser(self.config['cookiefile']))

	def resumable(self, expiration=3600):
		"""
		"""
		resume = False
		if 'cookiefile' in self.config and self.config['cookiefile'] is not None:
			if os.path.isfile(os.path.expanduser(self.config['cookiefile'])):
				self.logger.debug("Attempting to use cookiefile %s", self.config['cookiefile'])
				try:
					cookiejar = cookielib.MozillaCookieJar(os.path.expanduser(self.config['cookiefile']))
					cookiejar.load()

					cookies = 0
					resume = True # everything loads, but lets check more!
					for index, cookie in enumerate(cookiejar):
						cookies = cookies + 1
						if int(cookie.expires) <= time.time():
							self.logger.info("%s expired %ss ago", cookie, int(cookie.expires - time.time()))
							resume = False
						else:
							self.logger.debug("%s will expire in %ss", cookie, int(cookie.expires - time.time()))
							cookie.expires = int(time.time() + expiration)
					if cookies == 0:
						resume = False

					if not resume:
						cookiejar.clear()

					cookiejar.save()
				except IOError:
					return False
		return resume

	def connect(self):
		"""
		Connects to mymeteor.ie, does a check to see if meteors up and working.
		"""
		if self.resumable():
			self.logger.info("Resuming from connect.")
			return True

		handle = self.get('https://www.mymeteor.ie/')
		soup = BeautifulSoup(handle)

		if soup.html.head.title.string == "Meteor: MyMeteor Login":
			self.logger.info("connect has correct HTML title.")
			return True
		return False

	def login(self, username, password):
		"""
		Login to mymeteor.ie

		Returns true if successful or false if fails.
		"""
		if self.resumable():
			self.logger.info("Resuming from login.")
			return True
		else:
			self.logger.info("Unable to resume, running connect from login.")
			self.connect()

		post = [
			('username', username),
			('userpass', password),
			('returnTo', '/')
		]
		handle = self.post('https://www.mymeteor.ie/go/mymeteor-login-manager', post)
		soup = BeautifulSoup(handle)

		# TODO: Might be billpay, atlas says the title is different
		if soup.html.head.title.string == "Meteor: Prepay Functions":
			self.logger.info("login has correct HTML title.")
			return True
		return False

	def message(self, phonenumber, message):
		"""
		Sends a message to the recipient.
		"""
		params = [
			('event', "smsAjax"),
			('func', "addEnteredMsisdns")
		]
		post = [
			('ajaxRequest', "addEnteredMSISDNs"),
			('remove', "-"),
			('add', "0|" + phonenumber)
		]
		handle = self.post('https://www.mymeteor.ie/mymeteorapi/index.cfm', post, params)
		
		params = [
			('event', "smsAjax"),
			('func', "sendSMS")
		]
		post = [
			('ajaxRequest', "sendSMS"),
			('messageText', message)
		]
		handle = self.post('https://www.mymeteor.ie/mymeteorapi/index.cfm', post, params)

		messageSentPattern = re.compile('showEl\(\"sentTrue\"\)', re.IGNORECASE)
		for result in messageSentPattern.findall(handle):
			return True
		return False

	def import_contacts(self):
		"""
		Returns the list of contacts within the phonebook.

		If it fails, there is no checks so it returns an empty list.
		"""
		params = [
			('event', "smsAjax"),
			('func', "initFwtPhonebook"),
			('ajaxRequest', "initFwtPhonebook")
		]
		handle = self.get('https://www.mymeteor.ie/mymeteorapi/index.cfm', params)

		params = [
			('event', "smsAjax"),
			('func', "searchFWTPhonebook"),
			('ajaxRequest', "searchFwtPhonebook"),
			('searchValue', "All:single")
		]
		handle = self.get('https://www.mymeteor.ie/mymeteorapi/index.cfm', params)

		phonePattern = re.compile('\"Add\"\,\"(.+)"\,\"(08\d{8})\"\,\"(.+)\"\,\"(.+)\"', re.IGNORECASE)
		phonebook = []
		for result in phonePattern.findall(handle):
			phonebook.append({'phonenumber': result[1], 'username': result[0]})
		return phonebook

